example: ipsec_api: enhance runtime scripts

asasidharan opened the pull request:
The network interfaces and configurations required for
the application changes based on the application mode in
which the test is being run. Introduce env variable
IPSEC_APP_MODE which will be set by run script.

The supported modes are:
0 - stand alone mode, 1 - live mode and 2 - router mode.

Though linux-gen supports multiple "loop" devices like loop1
and loop2, other ODP implementations may not support
multiple loop interfaces. To allow flexibility for other
platforms, introduce following env variables in the runtime
scripts for standalone mode:
* IF_LIST env variable lists all the interfaces used for
  standalone mode of test.
* For outbound tests, stream entry has source interface as IN_IF
  and destination interface as OUT_IF.
* For inbound tests, stream entry has source interfaces as OUT_IF
  and destination interfaces as IN_IF.
* Introduce ROUTE_IF_INB and ROUTE_IF_OUTB for routing the IPSEC
  processed packets in standalone tests.

Introduce platform specific pktio_env script to define these
environment variables.

Signed-off-by: Aakash Sasidharan <asasidharan@marvell.com>
Change-Id: I656bb15d6f32ff166a6391dfed9f41ee2408591b

asasidharan commented:
@JannePeltonen could you please review this PR. This PR is for enhancing the run time scripts in ipsec_api application to make it a bit more CI friendly.

MatiasElo commented on example/ipsec_api/odp_ipsec_api_run_router.sh:
```diff
@@ -3,7 +3,55 @@
+PID=app_pid
+
+($STDBUF \
+ ./odp_ipsec_api -i $IF0,$IF2 \
+	-r 192.168.111.2/32,$IF0,$MAC_ADDR_IF1 \
+	-r 192.168.222.2/32,$IF2,$MAC_ADDR_IF3 \
```
Same comments as for `example/ipsec_api/odp_ipsec_api_run_live.sh`.

asasidharan replied:
Ack.

MatiasElo reviewed (commented):
Some minor comments mainly related to variable naming. I'd like to minimize the number of variables and keep the naming consistent among all the tests scripts if possible.

I did a quick test run and the scripts `odp_ipsec_api_run_esp_in.sh` and `odp_ipsec_api_run_esp_tun_in.sh` don't seem to pass and run indefinitely. However, this is unrelated to the changes done in this PR.

asasidharan commented on example/ipsec_api/odp_ipsec_api_run_ah_in.sh:
```diff
@@ -4,9 +4,36 @@
--s 192.168.222.2,192.168.111.2,loop2,loop1,10,100 \
--c 2 "$@"
+
+# IPSEC_APP_MODE: 0 - STAND ALONE, 1 - LIVE, 2 - ROUTER
```
Ack. Will update in the next revision.

asasidharan replied:
> Should be written: `STANDALONE`

Updated to "STANDALONE" with v6.

> 
> 
> > This behavior is observed even with the existing scripts. I believe this is an application bug. I did open a github issue earlier related to this: #1167.
> 
> Fix submitted in PR #1283.

Thanks @JannePeltonen for the fix.

asasidharan commented on example/ipsec_api/odp_ipsec_api_run_ah_in.sh:
```diff
@@ -4,9 +4,36 @@
+
+setup_interfaces
+
+./odp_ipsec_api -i $IF_LIST \
```
IF_LIST variable is required for our platform.
In our environment, IN_IF and OUT_IF refers to the same pktio interface so replacing IF_LIST with IN_IF and OUT_IF is not an option.

MatiasElo replied:
OK

asasidharan commented on example/ipsec_api/odp_ipsec_api_run_live.sh:
```diff
@@ -3,13 +3,59 @@
+PID=app_pid
+
+($STDBUF \
+ ./odp_ipsec_api -i $IF0,$IF2 \
+	-r 192.168.111.2/32,$IF0,$MAC_ADDR_IF1 \
+	-r 192.168.222.2/32,$IF2,$MAC_ADDR_IF3 \
```
IN_IF and OUT_IF is only used for the stream option to specify the source and destination interfaces while generating packet stream from application. This applies only for standalone mode.

For live and router mode, we don;t need a stream option as both the modes tests actual traffic generated with ping.

> The MAC address variable naming could also be more descriptive. E.g. NEXT_HOP_MAC0

Agreed. Will update so in the next revision.

MatiasElo replied:
OK, in this case more logical naming would be `IF0` & `IF1` and `NEXT_HOP_MAC0` & `NEXT_HOP_MAC1`. Same goes for `odp_ipsec_api_run_router.sh`.

asasidharan replied:
Ack. For live and router mode scripts, updated env variables to IF0 & IF1 and NEXT_HOP_MAC0 & NEXT_HOP_MAC1.

asasidharan commented:
> I did a quick test run and the scripts `odp_ipsec_api_run_esp_in.sh` and `odp_ipsec_api_run_esp_tun_in.sh` don't seem to pass and run indefinitely.

This behavior is observed even with the existing scripts. I believe this is an application bug. I did open a github issue earlier related to this: https://github.com/OpenDataPlane/odp/issues/1167.

MatiasElo commented on example/ipsec_api/Makefile.am:
```diff
@@ -2,18 +2,24 @@ include $(top_srcdir)/example/Makefile.inc
+if test_example
+TESTS = \
+	odp_ipsec_api_run_esp_out.sh \
+	odp_ipsec_api_run_simple.sh
```
Since these tests are really short you could add most of them here to increase CI testing coverage.

There seems to be something wrong with `odp_ipsec_api_run_esp_in.sh` and `odp_ipsec_api_run_esp_tun_in.sh` tests as they run indefinitely, but that's unrelated to this PR.

asasidharan replied:
Ack.

In v7, enabling all the tests except:
* ESP inbound tests
* Live an Router mode tests

MatiasElo commented:
> > I did a quick test run and the scripts `odp_ipsec_api_run_esp_in.sh` and `odp_ipsec_api_run_esp_tun_in.sh` don't seem to pass and run indefinitely.
> 
> This behavior is observed even with the existing scripts. I believe this is an application bug. I did open a github issue earlier related to this: #1167.

Ok

JannePeltonen commented:
> This behavior is observed even with the existing scripts. I believe this is an application bug. I did open a github issue earlier related to this: #1167.

Fix submitted in PR #1283.

MatiasElo reviewed (approved):
Looks good.

One typo (extra s) in the first commit message. Should be:
`The network interfaces and configurations required for
the application change based on...`

In the second commit message I'd say:
`ipsec_api_example.sh script is now unnecessary and can be removed.`

asasidharan commented:
With v9:
- Updated example/ipsec_api/Makefile.am to use relative path to pktio_env script.
- Updated the commit messages as suggested.
- Added reveiwed-by tag to both the commits.
